File: leetcode/p0074/Solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = 0;
        
        for (int i = 0; i < matrix.length; ++i) {
            if (matrix[i][0] > target) {
                break;
            } 
            row = i;
        }
        
        for (int i = 0; i < matrix[row].length; ++i) {
            if (matrix[row][i] == target) {
                return true;
            } else if (matrix[row][i] > target) {
                return false;
            }
        }
        return false;
    }
}

File: leetcode/p0074/Solution.kt
class Solution {
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        val col = matrix[0].size
        var left = 0
        var right = matrix.size * col - 1
        while (left <= right) {
            val mid = (left + right) ushr 1
            val num = matrix[mid / col][mid % col]
            if (num == target) {
                return true
            } else if (num < target) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        return false
    }
}
